polkadot/node/network/collator-protocol/src/validator_side_experimental/peer_manager/mod.rs

```diff
@@ -0,0 +1,516 @@
+			"Disconnecting peers",
+		);
+
+		sender
+			.send_message(NetworkBridgeTxMessage::DisconnectPeers(
+				peers.into_iter().collect(),
+				PeerSet::Collation,
+			))
+			.await;
+	}
+}
+
+async fn get_ancestors<Sender: CollatorProtocolSenderTrait>(
+	sender: &mut Sender,
+	k: usize,
+	hash: Hash,
+) -> Result<Vec<Hash>> {
+	let (tx, rx) = oneshot::channel();
+	sender
+		.send_message(ChainApiMessage::Ancestors { hash, k, response_channel: tx })
+		.await;
+
+	Ok(rx.await.map_err(|_| Error::CanceledAncestors)??)
+}
+
+async fn get_latest_finalized_block<Sender: CollatorProtocolSenderTrait>(
+	sender: &mut Sender,
+) -> Result<(BlockNumber, Hash)> {
+	let (tx, rx) = oneshot::channel();
+	sender.send_message(ChainApiMessage::FinalizedBlockNumber(tx)).await;
+
+	let block_number = rx.await.map_err(|_| Error::CanceledFinalizedBlockNumber)??;
+
+	let (tx, rx) = oneshot::channel();
+	sender.send_message(ChainApiMessage::FinalizedBlockHash(block_number, tx)).await;
+
+	let block_hash = rx
+		.await
+		.map_err(|_| Error::CanceledFinalizedBlockHash)??
+		.ok_or_else(|| Error::FinalizedBlockNotFound(block_number))?;
+
+	Ok((block_number, block_hash))
+}
+
+async fn extract_reputation_bumps_on_new_finalized_block<Sender: CollatorProtocolSenderTrait>(
+	sender: &mut Sender,
+	processed_finalized_block_number: BlockNumber,
+	(latest_finalized_block_number, latest_finalized_block_hash): (BlockNumber, Hash),
+) -> Result<BTreeMap<ParaId, HashMap<PeerId, Score>>> {
+	if latest_finalized_block_number < processed_finalized_block_number {
+		// Shouldn't be possible, but in this case there is no other initialisation needed.
+		gum::info!(
+			target: LOG_TARGET,
+			latest_finalized_block_number,
+			?latest_finalized_block_hash,
+			"Peer manager stored finalized block number {} is higher than the latest finalized block.",
+			processed_finalized_block_number,
+		);
+		return Ok(BTreeMap::new())
+	}
+
+	let ancestry_len = std::cmp::min(
+		latest_finalized_block_number.saturating_sub(processed_finalized_block_number),
+		MAX_STARTUP_ANCESTRY_LOOKBACK,
+	);
+
+	if ancestry_len == 0 {
+		return Ok(BTreeMap::new())
+	}
+
+	let mut ancestors =
+		get_ancestors(sender, ancestry_len as usize, latest_finalized_block_hash).await?;
+	ancestors.push(latest_finalized_block_hash);
+	ancestors.reverse();
+
+	gum::trace!(
+		target: LOG_TARGET,
+		?latest_finalized_block_hash,
+		processed_finalized_block_number,
+		"Processing reputation bumps for finalized relay parent {} and its {} ancestors",
+		latest_finalized_block_number,
+		ancestry_len
+	);
+
+	let mut candidates_per_rp: HashMap<Hash, BTreeMap<ParaId, HashSet<CandidateHash>>> =
+		HashMap::with_capacity(ancestors.len());
+
+	for i in 1..ancestors.len() {
+		let rp = ancestors[i];
+		let parent_rp = ancestors[i - 1];
+		let candidate_events = recv_runtime(request_candidate_events(rp, sender).await).await?;
+
+		for event in candidate_events {
+			if let CandidateEvent::CandidateIncluded(receipt, _, _, _) = event {
```

Comment:
nit: We can completely avoid calling `expensive request_candidates_pending_availability()` for paras that have no `v2 receipts backed`.

Reply:
good point